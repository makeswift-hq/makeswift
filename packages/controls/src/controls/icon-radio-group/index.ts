export * from './icon-radio-group'
