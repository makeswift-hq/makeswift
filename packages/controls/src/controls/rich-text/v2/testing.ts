import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
