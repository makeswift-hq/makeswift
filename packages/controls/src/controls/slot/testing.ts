import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
