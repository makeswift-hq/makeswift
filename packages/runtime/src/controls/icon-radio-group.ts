export const unstable_IconRadioGroupIcon = {
  TextAlignCenter: 'TextAlignCenter',
  TextAlignJustify: 'TextAlignJustify',
  TextAlignLeft: 'TextAlignLeft',
  TextAlignRight: 'TextAlignRight',
  Superscript: 'Superscript16',
  Subscript: 'Subscript16',
  Code: 'Code16',
} as const

export type IconRadioGroupIcon =
  typeof unstable_IconRadioGroupIcon[keyof typeof unstable_IconRadioGroupIcon]

export const IconRadioGroupControlType = 'makeswift::controls::icon-radio-group'

export type IconRadioGroupOption<T extends string> = {
  value: T
  icon: IconRadioGroupIcon
  label: string
}

type IconRadioGroupControlConfig<T extends string = string> = {
  label?: string
  defaultValue?: T
  options: IconRadioGroupOption<T>[]
}

export type IconRadioGroupControlDefinition<
  C extends IconRadioGroupControlConfig = IconRadioGroupControlConfig,
> = {
  type: typeof IconRadioGroupControlType
  config: C
}

export type IconRadioGroupControlDefinitionOption<
  T extends IconRadioGroupControlDefinition = IconRadioGroupControlDefinition,
> = T['config'] extends IconRadioGroupControlConfig<infer U> ? U : never

export type IconRadioGroupControlData<
  T extends IconRadioGroupControlDefinition = IconRadioGroupControlDefinition,
> = IconRadioGroupControlDefinitionOption<T>

export function unstable_IconRadioGroup<T extends string, C extends IconRadioGroupControlConfig<T>>(
  config: C & { options: IconRadioGroupOption<T>[] },
): IconRadioGroupControlDefinition<C> {
  return { type: IconRadioGroupControlType, config }
}
