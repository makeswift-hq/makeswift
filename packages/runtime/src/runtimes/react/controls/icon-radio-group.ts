import { IconRadioGroupControlData, IconRadioGroupControlDefinition } from '@makeswift/controls'

export type IconRadioGroupControlValue<T extends IconRadioGroupControlDefinition> =
  undefined extends T['config']['defaultValue']
    ? IconRadioGroupControlData | undefined
    : IconRadioGroupControlData

export function useIconRadioGroupValue<T extends IconRadioGroupControlDefinition>(
  iconRadioGroupControlData: IconRadioGroupControlData | undefined,
  controlDefinition: T,
): IconRadioGroupControlValue<T> {
  return (iconRadioGroupControlData ??
    controlDefinition.config.defaultValue) as IconRadioGroupControlValue<T>
}
