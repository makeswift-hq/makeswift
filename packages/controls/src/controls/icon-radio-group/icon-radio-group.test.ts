import { testDefinition, testResolveValue } from '../../testing/test-definition'

import { unstable_IconRadioGroup as IconRadioGroup } from './icon-radio-group'

const options = [
  {
    value: 'code',
    label: 'Code',
    icon: IconRadioGroup.Icon.Code,
  },
  {
    value: 'superscript',
    label: 'Superscript',
    icon: IconRadioGroup.Icon.Superscript,
  },
] as const

describe('unstable_IconRadioGroup', () => {
  describe('constructor', () => {
    test('returns correct definition', () => {
      const def = IconRadioGroup({
        label: 'Block type',
        options,
        defaultValue: 'code',
      })

      expect(def).toMatchSnapshot()
    })

    test('enforces item value type', () => {
      IconRadioGroup({
        label: 'Block type',
        options: [
          {
            // @ts-expect-error
            value: 42,
            label: 'Code',
            icon: IconRadioGroup.Icon.Code,
          },
        ],
      })
    })

    test('disallows extraneous properties', () => {
      IconRadioGroup({
        options,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    // FIXME
    // function assignTest(_def: IconRadioGroupDefinition) {}
    // assignTest(IconRadioGroup({ options }))
    // assignTest(IconRadioGroup({ label: 'Block type', options }))
    // assignTest(
    //   IconRadioGroup({ label: 'Block type', options, defaultValue: 'code' }),
    // )
    // assignTest(IconRadioGroup({ label: 'visible', defaultValue: undefined }))
    // assignTest(IconRadioGroup({ label: undefined, defaultValue: undefined }))
  })
})

describe.each([
  [IconRadioGroup({ options }), ['code', 'superscript', undefined] as const],
  [
    IconRadioGroup({ options, defaultValue: 'code' }),
    ['code', 'superscript'] as const,
  ],
])('unstable_IconRadioGroup', (def, values) => {
  const invalidValues = [null, 17, 'random', { swatchId: 42 }]
  testDefinition(def, values, invalidValues)
  testResolveValue(def, values)
})
