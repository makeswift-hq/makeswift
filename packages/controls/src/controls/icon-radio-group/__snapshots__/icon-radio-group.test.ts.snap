// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unstable_IconRadioGroup constructor returns correct definition 1`] = `
IconRadioGroupDefinition {
  "config": {
    "defaultValue": "code",
    "label": "Block type",
    "options": [
      {
        "icon": "Code16",
        "label": "Code",
        "value": "code",
      },
      {
        "icon": "Superscript16",
        "label": "Superscript",
        "value": "superscript",
      },
    ],
  },
}
`;

exports[`unstable_IconRadioGroup definition w/ config {"label":"Block type","options":[{"value":"code","label":"Code","icon":"Code16"},{"value":"superscript","label":"Superscript","icon":"Superscript16"}],"defaultValue":"code"} toData returns v1 data for \`code\` when definition version is 1 1`] = `"code"`;

exports[`unstable_IconRadioGroup definition w/ config {"label":"Block type","options":[{"value":"code","label":"Code","icon":"Code16"},{"value":"superscript","label":"Superscript","icon":"Superscript16"}],"defaultValue":"code"} toData returns v1 data for \`superscript\` when definition version is 1 1`] = `"superscript"`;
